### PR TITLE
Build the ToolSystem that include weapon, tools, equipment e.d.

### PR DESCRIPTION
**Pull request by onon1101**
first refactor the arch that add coin in the player package

### CODE REVIEW

File: include/Player/Equipment/Weapon/Component.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    int16_t golden;
+    std::string obsidian;
+    int16_t titanium;
+};
+
+// Settings
+struct Settings {
+    explicit Settings(std::string &path, bool isThrow)
+        : imagePath(path),
+          isThrow(isThrow){};
```
Comment (HeavenManySugar) [inline include/Player/Equipment/Weapon/Component.h:38]: warning: field 'imagePath' will be initialized after field 'isThrow'

File: include/Player/Equipment/IEquipment.h [hunk 2]
```diff
@@ -8,14 +8,24 @@
...
-    virtual std::shared_ptr<GameElement> getGameElement() = 0;
+    enum Direction {
+        ROW = 0,
+        COL,
+    };
 
+    // setter
     virtual void setPosition(const glm::vec2 position) = 0;
+    virtual void setWeaponType(WeaponEnum::Type type) { return;};
+    virtual void setShovelType(ShovelEnum::Type type) { return;};
```
Comment (HeavenManySugar) [inline include/Player/Equipment/IEquipment.h:23]:
```cpp
    virtual void setWeaponType([[maybe_unused]] WeaponEnum::Type type) { return;};
    virtual void setShovelType([[maybe_unused]] ShovelEnum::Type type) { return;};
```

File: src/Player/Equipment/Weapon.cpp [hunk 3]
```diff
@@ -2,44 +2,71 @@
...
 
-    const auto ItemImage = std::make_shared<SpriteSheet>(
-        m_ImagePathItem, ItemSize, std::vector<std::size_t>{0}, false, 100, true, 100
-    );
-    const auto WindowImage = std::make_shared<Util::Image>(m_ImagePathWindow);
+    rendererWindow();
 
-    m_Window->SetDrawable(WindowImage);
-    m_Item->SetDrawable(ItemImage);
+    setWeaponType(WeaponEnum::Type::DAGGER);
```
Comment (HeavenManySugar) [inline src/Player/Equipment/Weapon.cpp:11]: image

image

**Review by HeavenManySugar — CHANGES_REQUESTED** [review]
Not work on my machine

File: include/Player/Equipment/Weapon/Component.h (hunk 1 above)
Comment (HeavenManySugar) [reply include/Player/Equipment/Weapon/Component.h:38]:
```cpp
struct Settings {
    explicit Settings(std::string &path, bool isThrow)
        : imagePath(path),
          isThrow(isThrow){};

    std::string imagePath;
    bool isThrow;
};
```

File: src/Player/Equipment/Weapon.cpp (hunk 3 above)
Comment (HeavenManySugar) [reply src/Player/Equipment/Weapon.cpp:11]: Do not call overridable methods in constructors
Reference: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2214

**Comment by onon1101** [comment]
please give some detail.

**Comment by HeavenManySugar** [comment]
帶有override修飾的function不能在建構子中被呼叫

File: src/Player/Items/Tools.cpp [hunk 4]
```diff
@@ -7,48 +7,116 @@
...
+        auto pos = m_colPosList[i];
+
+        equipList->setPosition(pos);
     }
-    for (size_t i = 0; i < b.size(); i++) {
-        m_TempGameObejct.push_back(b[i]);
+}
+
+void Tools::rearrangeRow() {
+    for (int i = 0; i < m_rowEquipList.size() ;i++) {
```
Comment (HeavenManySugar) [inline src/Player/Items/Tools.cpp:102]:
```cpp
    for (size_t i = 0; i < m_rowEquipList.size() ;i++) {
```

File: src/Player/Items/Tools.cpp [hunk 5]
```diff
@@ -7,48 +7,116 @@
...
     }
-    for (size_t i = 0; i < t.size(); i++) {
-        m_TempGameObejct.push_back(t[i]);
+
+    m_Tools->SetVisible(false);
+    return m_Tools;
+}
+
+void Tools::rearrangeCol() {
+    for (int i = 0; i < m_colEquipList.size() ;i++) {
```
Comment (HeavenManySugar) [inline src/Player/Items/Tools.cpp:93]: for (size_t i = 0; i < m_colEquipList.size() ;i++) {

**Review by onon1101 — COMMENTED** [review]
ok

**Review by HeavenManySugar — APPROVED** [review]

**Pull request merged**